Alerting: Reduce set of fields that could trigger alert state change

benoittgt opened the pull request:
**What is this feature?**

We want to avoid too much change of alert state based on change on alert's fields. For that we ignore some fields from the diff.

This is a rather cautious first pull request, ~~but I would love to make another pull request for `IntervalSeconds`~~. We could even go further as [pointed](https://github.com/grafana/grafana/issues/83250) by @yuri-tceretian but I can do in another PR.

**Why do we need this feature?**

There is great issue by @yuri-tceretian  that detail all fo this https://github.com/grafana/grafana/issues/83250

**Who is this feature for?**

Everyone who use alerting and updates.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/83250

cc @yuri-tceretian

benoittgt committed:
Reduce set of fields that could trigger alert state change

We want to avoid too much change of alert state based on change on
alert's fields. For that we ignore some fields from the diff.

This is a rather cautious first pull request, but I would love to make
another pull request for `IntervalSeconds`.

Close: https://github.com/grafana/grafana/issues/83250

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=83496) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=83496) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=83496) it.</sub>

grobinson-grafana commented on pkg/services/ngalert/schedule/registry.go:
```diff
@@ -326,7 +308,6 @@ func (r ruleWithFolder) Fingerprint() fingerprint {
 	writeString(rule.Title)
 	writeString(rule.NamespaceUID)
 	writeString(r.folderTitle)
-	writeLabels(rule.Labels)
```
Looking at Yuri's comments, we cannot remove labels as it is fundamental to the fingerprint.

benoittgt replied:
You are correct thanks.

yuri-tceretian commented on pkg/services/ngalert/schedule/registry.go:
```diff
@@ -347,7 +328,6 @@ func (r ruleWithFolder) Fingerprint() fingerprint {
-	writeLabels(rule.Annotations)
 	if rule.DashboardUID != nil {
 		writeString(*rule.DashboardUID)
 	}
```
Only fields below //TODO are up for removal. Other fields cannot be removed.

benoittgt committed:
Revert "Reduce set of fields that could trigger alert state change"

This reverts commit 1dba97afac938439a356a0a913832d8aaa77ac0c.

benoittgt committed:
Remove 2 fields that could trigger alert state change

We want to avoid too much change of alert state based on change on
alert's fields. For that we ignore some fields from the diff.

Close: #83250

yuri-tceretian reviewed (approved):
LGTM. Thanks for the contribution! I really appreciate it!
We also can delete DashboardUID and PanelID because they are the product of annotations. Removal of ID could help us handle situations when a rule is deleted and a new one with the same UID is created. No need to address it here. We can open another PR.

benoittgt committed:
Handle failure with "assignment to entry in nil map"

yuri-tceretian commented on pkg/services/ngalert/schedule/registry_test.go:
```diff
@@ -369,10 +369,12 @@ func TestRuleWithFolderFingerprint(t *testing.T) {
 		cp.Version++
 		cp.Updated = cp.Updated.Add(1 * time.Second)
+		cp.IntervalSeconds++
+		cp.Annotations["test"] = "test"
```
The failure is here. Please cerate a new annotation map.

yuri-tceretian commented:
Unfortunately, I forgot to merge it before Grafana 11 cutoff. This will be included only in Grafana 11.1

benoittgt commented:
@yuri-tceretian no problem. Thanks for the merge. Do you know when 11.1 will be released?

ashharrison90 commented:
@benoittgt this actually did make it within the cutoff, so will be released with grafana 11 👍
